drivers/s3.go

```diff
@@ -0,0 +1,294 @@
+	s3svc      *s3.S3
+	manifestID string
+	nonce      uint64
+}
+
+// S3BUCKET s3 bucket owned by this node
+var S3BUCKET string
+
+// S3REGION region of s3 bucket owned by this node
+var S3REGION string
+
+// IsOwnStorageS3 returns true if turi points to S3 bucket owned by this node
+func IsOwnStorageS3(turi *net.TypedURI) bool {
+	if turi.Storage != "s3" || S3BUCKET == "" || S3REGION == "" {
+		return false
+	}
+	return strings.Contains(turi.Uri, S3BUCKET) && strings.Contains(turi.Uri, S3REGION)
+}
+
+func newS3Driver(info *net.S3OSInfo) OSDriver {
+	od := &s3OS{
+		region:         info.Region,
+		bucket:         info.Bucket,
+		policy:         info.Policy,
+		signature:      info.Signature,
+		xAmzDate:       info.XAmzDate,
+		xAmzCredential: info.XAmzCredential,
+	}
+	return od
+}
+
+func NewOwnS3Driver(region, bucket, accessKey, accessKeySecret string) OSDriver {
+	od := &s3OS{
+		region:             region,
+		bucket:             bucket,
+		awsAccessKeyID:     accessKey,
+		awsSecretAccessKey: accessKeySecret,
+	}
+	return od
+}
+
+func (os *s3OS) IsExternal() bool {
+	return true
+}
+
+func (os *s3OS) StartSession(jobID int64, manifestID string, nonce uint64) OSSession {
+	glog.V(common.DEBUG).Infof("S3 Starting session for job %d man %s nonce %d", jobID, manifestID, nonce)
+	s := s3OSSession{
+		osd:        os,
+		created:    time.Now(),
+		jobID:      jobID,
+		manifestID: manifestID,
+		nonce:      nonce,
+	}
+	if os.awsAccessKeyID != "" {
+		s.creds = credentials.NewStaticCredentials(os.awsAccessKeyID, os.awsSecretAccessKey, "")
+		cfg := aws.NewConfig().WithRegion(s.osd.region).WithCredentials(s.creds)
+		s.s3svc = s3.New(session.New(), cfg)
+	}
+	return &s
+}
+
+func (session *s3OSSession) EndSession() {
+
+}
+
+func (session *s3OSSession) IsOwnStorage(turi *net.TypedURI) bool {
+	return IsOwnStorageS3(turi)
+}
+
+func (session *s3OSSession) GetInfo() *net.OSInfo {
+	info := net.OSInfo{
+		Storage: "s3",
+		S3Info: &net.S3OSInfo{
+			Region: session.osd.region,
+			Bucket: session.osd.bucket,
+		},
+	}
+	if session.osd.awsAccessKeyID != "" {
+		policy, signature, xAmzCredential, xAmzDate := createPolicy(session.osd.awsAccessKeyID,
+			session.osd.bucket, session.osd.region, session.osd.awsSecretAccessKey)
+		info.S3Info.Policy = policy
+		info.S3Info.XAmzCredential = xAmzCredential
+		info.S3Info.XAmzDate = xAmzDate
+		info.S3Info.Signature = signature
+	}
+	return &info
+}
+
+func (session *s3OSSession) getAbsURL(path string) string {
+	return fmt.Sprintf("https://s3.%s.amazonaws.com/%s/%s", session.osd.region, session.osd.bucket, path)
+}
+
+func (session *s3OSSession) SaveData(streamID, name string, data []byte) (*net.TypedURI, string, error) {
+	glog.V(common.VERBOSE).Infof("saving data to s3 %s", name)
+	path := fmt.Sprintf("%s_%d", session.manifestID, session.nonce)
+	var err error
+	if session.s3svc == nil {
+		path, err = session.postData(path, name, data)
+	} else {
+		path, err = session.putData(path, name, data)
```

Comment:
We can, but using `post` require generation access policy - so it will be unnecessary calculations on each save.

Reply:
Can't we reuse the POST policy across requests for the duration of a job?

Reply:
I'm removing session class, so 'duration of the job' will not be tracked. Plus post policy has expiration time - so we will need to track that and issue new policy when previous one expired.
That's, btw, problem I left unsolved for now - T issues policy valid for 24 hours, and if broadcasting will be run longer than that without interruption - it will expire and B will not be able put save more data into T's S3

Reply:
> 'duration of the job' will not be tracked.

We already have a few notions of sessionability  --
* Broadcaster connection liveness
* Transcode loop liveness

For S3, the broadcaster and transcoder can create fresh POST policies and use the same policy throughout the session.

> Plus post policy has expiration time

The POST policy expiration can be an approximation of the time remaining in the job. Something like `secsToExpiration = (job.EndBlock - currentBlock) * 4`. A 24h default (eg, for offchain transcoding) is probably fine as well.